src/sliceDataStorage.cpp

```diff
@@ -469,4 +477,113 @@ bool SliceDataStorage::getExtruderPrimeBlobEnabled(int extruder_nr) const
     return train->getSettingBoolean("prime_blob_enable");
 }
 
+
+void SupportLayer::excludeAreasFromSupportInfillAreas(const Polygons& exclude_polygons, const AABB& exclude_polygons_boundary_box)
+{
+    // record the indexes that need to be removed and do that after
+    std::vector<size_t> to_remove_indexes;  // FIFO for replacing, LIFO for removing
+    to_remove_indexes.reserve(this->support_infill_parts.size());
+    size_t to_remove_indexes_front_idx = 0;
+
+    std::vector<PolygonsPart> new_smaller_islands;
+    new_smaller_islands.reserve(this->support_infill_parts.size());
+
+    for (size_t part_idx = 0; part_idx < this->support_infill_parts.size(); ++part_idx)
+    {
+        SupportInfillPart& support_infill_part = this->support_infill_parts[part_idx];
+
+        // if the areas don't overlap, do nothing
+        if (!exclude_polygons_boundary_box.hit(support_infill_part.outline_boundary_box))
+        {
+            continue;
+        }
+
+        Polygons result_polygons = support_infill_part.outline.difference(exclude_polygons);
+        std::vector<PolygonsPart> smaller_support_islands = result_polygons.splitIntoParts();
+
+        // if no smaller parts get generated, this mean this part should be removed.
+        // instead of directly removing it, we replace it with a pending newly generated part if there is one.
+        if (smaller_support_islands.empty())
+        {
+            if (new_smaller_islands.empty())
+            {
+                // there is no islands to add, append this into the to_remove list
+                to_remove_indexes.push_back(part_idx);
+            }
+            else
+            {
+                // replace the current part with a newly generated part
+                support_infill_part.outline = new_smaller_islands.back();
+                new_smaller_islands.pop_back();
+            }
+            continue;
+        }
+
+        // there are one or more smaller parts.
+        // we first replace the current part with one of the smaller parts, for the rest, we keep replacing the
+        // parts that need to be removed until there is none, and then we put the rest into the newly generated
+        // parts list.
+        support_infill_part.outline = smaller_support_islands[0];
+
+        // reserve space beforehand
+        if (new_smaller_islands.capacity() < new_smaller_islands.size() + smaller_support_islands.size())
```

Comment:
In the beginning of the function I did a `new_smaller_islands.reserve(this->support_infill_parts.size());` which I think should work in most cases. Because `this->support_infill_parts.size()` is probably more than what we need, we probably over-allocated in the beginning and don't need to reserve more again at all.

This check is only used in case we generate way more smaller support islands than we have expected, and in that case we want to reserve more. but in other cases, we don't want to downsize it. That's why I used the `capacity` instead of size to do the check.

This is probably over-engineered and let's discuss next week in person. I think that would be better :)